invoiceregistry.go

```diff
@@ -51,7 +59,123 @@ func newInvoiceRegistry(cdb *channeldb.DB) *invoiceRegistry {
 		cdb:                 cdb,
 		debugInvoices:       make(map[chainhash.Hash]*channeldb.Invoice),
 		notificationClients: make(map[uint32]*invoiceSubscription),
+		newSubscriptions:    make(chan *invoiceSubscription),
+		subscriptionCancels: make(chan uint32),
+		invoiceEvents:       make(chan *invoiceEvent),
+		quit:                make(chan struct{}),
+	}
+}
+
+// Start starts the registry and all goroutines it needs to carry out its task.
+func (i *invoiceRegistry) Start() error {
+	i.wg.Add(1)
+
+	go i.invoiceEventNotifier()
+
+	return nil
+}
+
+// Stop signals the registry for a graceful shutdown.
+func (i *invoiceRegistry) Stop() {
+	close(i.quit)
+
+	i.wg.Wait()
+}
+
+// invoiceEvent represents a new event that has modified on invoice on disk.
+// Only two event types are currently supported: newly created invoices, and
+// instance where invoices are settled.
+type invoiceEvent struct {
+	isSettle bool
+
+	invoice *channeldb.Invoice
+}
+
+// invoiceEventNotifier is the dedicated goroutine responsible for accepting
+// new notification subscriptions, cancelling old subscriptions, and
+// dispatching new invoice events.
+func (i *invoiceRegistry) invoiceEventNotifier() {
+	defer i.wg.Done()
+
+	for {
+		select {
+		// A new invoice subscription has just arrived! We'll query for
+		// any backlog notifications, then add it to the set of
+		// clients.
+		case newClient := <-i.newSubscriptions:
+			// Before we add the client to our set of active
+			// clients, we'll first attempt to deliver any backlog
+			// invoice events.
+			err := i.deliverBacklogEvents(newClient)
+			if err != nil {
+				ltndLog.Errorf("unable to deliver backlog invoice "+
+					"notifications: %v", err)
+			}
+
```

Comment:
maybe need a check to ensure the next add and settle indexes are aligned with upcoming ntfns

Reply:
I think this can be done by keeping track of the "last delivered" add/settle, then passing these into `deliverBacklogEvents`. If it delivers up to and including those values, the next things in the pipeline should be gtg

Reply:
What do you mean aligned with the upcoming notifications? To ensure that there isn't a gap between what the client new last, and the next upcoming notification? 

The aspiration here was that since we deliver these backlogs within the main event loop, the client _should_ then get all the notifications since that point as well.

Reply:
right now i believe it'd be possible to deliver duplicate notifications since the database state can be modified outside of the main loop.

to prevent this, we should clamp the delivery of historical adds/settles using the last delivered heights of each